netty/src/main/java/io/grpc/transport/netty/Http2Negotiator.java

```diff
@@ -112,168 +97,252 @@ public static ChannelHandler serverTls(SSLEngine sslEngine) {
-    final ChannelInitializer<Channel> initializer = new ChannelInitializer<Channel>() {
-      @Override
-      public void initChannel(Channel ch) throws Exception {
-        ch.pipeline().addLast(upgrader);
-        ch.pipeline().addLast(completionHandler);
+    return new BufferUntilHttpUpgradeCompleteHandler(upgrader);
+  }
+
+  /**
+   * Returns a {@link ChannelHandler} that ensures that the {@code handler} is added to the
+   * pipeline writes to the {@link Channel} may happen immediately, even before it is active.
+   */
+  public static ChannelHandler plaintext(final ChannelHandler handler) {
+    return new BufferUntilChannelActiveHandler(handler);
+  }
+
+  /**
+   * Buffers all writes until either {@link #flushBufferedAndRemove(ChannelHandlerContext)} or
+   * {@link #failBufferedAndClose(ChannelHandlerContext)} is called. This handler allows us to
+   * write to a {@link Channel} before we are allowed to write to it officially i.e.
+   * before it's active or the TLS Handshake is complete.
+   */
+  private abstract static class BufferWritesHandler extends ChannelDuplexHandler {
+
+    private ChannelHandler[] handlers;
+    private Queue<ChannelWrite> bufferedWrites = new ArrayDeque<ChannelWrite>();
+    private boolean flushed;
+
+    /**
+     * @param handlers the ChannelHandlers are added to the pipeline on channelRegistered and
+     *                 before this handler.
+     */
+    BufferWritesHandler(ChannelHandler... handlers) {
+      this.handlers = handlers;
+    }
+
+    @Override
+    public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
+      /**
+       * This check is necessary as a channel may be registered with different eventloops during
+       * its lifetime and we only want to set up the pipeline once.
+       * We can't set up the pipeline by specifying a ChannelInitializer in the client bootstrap,
+       * because the initializer is not executed in the calling thread but on channelRegistered
+       * in the eventloop thread. So there is a race condition between threads immediately writing
+       * to the channel (from outside the eventloop) and the pipeline being set up i.e. a
+       * write task (writing to the HeadContext) may be created before the pipeline is set up.
+       */
+      if (handlers != null) {
+        ctx.pipeline().addFirst(handlers);
+        handlers = null;
       }
-    };
+      super.channelRegistered(ctx);
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+      failBufferedAndClose(ctx);
+      super.channelInactive(ctx);
+    }
 
-    return new Negotiation() {
-      @Override
-      public ChannelInitializer<Channel> initializer() {
-        return initializer;
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
+        throws Exception {
+      /**
+       * This check is necessary as there is a race condition between threads writing to the
```

Comment:
This comment is a little confusing to me.  I think you meant something like this:

"This check is due to a race condition between writing and removal of this handler. When this handler removes itself, the removal event is put at the end of the EventLoop's queue, which may be preceded by additional write requests. To handle this case, we mark removal of this handler by setting bufferedWrites = null so that any additional queued writes are just passed on to the next handler in the pipeline."